File: 2018/day05-alchemic-reduction/polymers.py
#!/usr/bin/env python

import time


def main() -> None:
    start = time.perf_counter()
    with open('day5.in') as f:
        polymer = f.read().strip()
    print(f"Part 1: {len(triggerReaction(polymer))}")
    print(f"Part 2: {len(shortestPolymer(polymer))}")
    print(f"Time taken: {time.perf_counter()-start} sec")


def triggerReaction(polymer: str) -> str:
    units, i = list(polymer), 0
    while i < len(units)-2:
        if abs(ord(units[i])-ord(units[i+1])) == 32:
            units = units[:i] + units[i+2:]
            if i:
                i -= 1
        else:
            i += 1
    return ''.join(units)


def shortestPolymer(polymer: str) -> str:
    table = {}
    for ch in polymer:
        if ch not in table:
            copy = polymer[:]
            table[ch] = triggerReaction(copy.replace(ch.lower(), "").
                                        replace(ch.upper(), ""))
    return min(table.values(), key=len)


if __name__ == '__main__':
    main()

File: 2018/day05-alchemic-reduction/polymers.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

func main() {
	start := time.Now()
	polymer := readPolymer("day5.in")
	fmt.Println("Part 1:", len(triggerReaction(polymer)))
	fmt.Println("Part 2:", len(shortestPolymer(polymer)))
	fmt.Println("Time taken:", time.Since(start).Seconds(), "sec")
}

func readPolymer(filename string) string {
	file, err := os.Open(filename)
	if err != nil {
		log.Fatal(err)
	}
	var polymer string
	sc := bufio.NewScanner(file)
	for sc.Scan() {
		polymer = sc.Text()
	}
	return polymer
}

// Part 1
func triggerReaction(polymer string) string {
	units := []rune(polymer)
	for i := 0; i < len(units)-1; {
		if sameTypeOppositePolarity(units[i], units[i+1]) {
			units = append(units[:i], units[i+2:]...)
			if i != 0 {
				i--
			}
		} else {
			i++
		}
	}
	return string(units)
}

// Same letter, different case e.g. ('a', 'A'), ('C', 'c')
func sameTypeOppositePolarity(char1, char2 int32) bool {
	return char1-char2 == 32 || char2-char1 == 32
}

// Part 2
func shortestPolymer(polymer string) string {
	minPolymer, table := polymer, make(map[string]string)
	for _, char := range polymer {
		ch := strings.ToLower(string(char))
		if _, ok := table[ch]; !ok {
			table[ch] = triggerReaction(removeAll(polymer, char))
			if len(table[ch]) < len(minPolymer) {
				minPolymer = table[ch]
			}
		}
	}
	return minPolymer
}

// Remove all uppercase and lowercase occurrences of char from polymer
func removeAll(polymer string, char rune) string {
	newPolymer := strings.ReplaceAll(polymer, strings.ToLower(string(char)), "")
	newPolymer = strings.ReplaceAll(newPolymer, strings.ToUpper(string(char)), "")
	return newPolymer
}
